### PR TITLE
[REF] mail: canned responses

### PR DESCRIPTION
**Pull request by jgi-odoo**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by jgi-odoo** [comment]
@alexkuhn @seb-odoo

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 1]
```diff
@@ -206,8 +412,8 @@ QUnit.test('display partner mention suggestions on typing "@"', async function (
         '.o_ComposerTextInput_mentionDropdownPropositionList',
         "mention suggestions list should not be present"
     );
+    document.querySelector(`.o_ComposerTextInput_textarea`).focus();
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer/composer_tests.js:415]: revert

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 2]
```diff
@@ -252,11 +458,9 @@ QUnit.test('mention a partner', async function (assert) {
         "",
         "text content of composer should be empty initially"
     );
+    document.querySelector(`.o_ComposerTextInput_textarea`).focus();
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer/composer_tests.js:461]: revert

File: addons/mail/static/src/components/composer/composer_tests.js [hunk 3]
```diff
@@ -372,11 +574,9 @@ QUnit.test('add an emoji after a partner mention', async function (assert) {
...
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer/composer_tests.js:577]: revert (looks better `.focus()` in `afterNextRender` inner function)

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js [hunk 4]
```diff
@@ -5,6 +5,7 @@ const useStore = require('mail/static/src/component_hooks/use_store/use_store.js
 
 const components = {
     PartnerMentionSuggestion: require('mail/static/src/components/partner_mention_suggestion/partner_mention_suggestion.js'),
+    CannedResponseSuggestion: require('mail/static/src/components/canned_response_suggestion/canned_response_suggestion.js'),
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer_text_input/composer_text_input.js:8]: alphabetical order

File: addons/mail/static/src/models/composer/composer.js [hunk 5]
```diff
@@ -228,6 +277,27 @@ function factory(dependencies) {
...
+                this.update({
+                    activeSuggestedCannedResponse: [[
+                        'link',
+                        this.suggestedCannedResponses[activeSuggestedCannedResponseIndex + 1]
+                    ]],
+                });
+            } else {
+                this.update({
+                    activeSuggestedCannedResponse: [['link', this.suggestedCannedResponses[0]]],
+                });
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:297]: Maybe it's easier to read with single `this.update()` with ternary expression?

File: addons/mail/static/src/models/composer/composer.js [hunk 6]
```diff
@@ -357,6 +463,15 @@ function factory(dependencies) {
             }
         }
 
+        _detectDelimiterCannedResponse() {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:466]: missing doc

File: addons/mail/static/src/models/composer/composer.js [hunk 7]
```diff
@@ -473,10 +588,32 @@ function factory(dependencies) {
             }
         }
 
+        _getSuggestedCannedResponses(mentionKeyword) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:591]: Missing doc, and it doesn't look like a "get" method at a glance.

File: addons/mail/static/src/models/composer/composer.js [hunk 8]
```diff
@@ -473,10 +588,32 @@ function factory(dependencies) {
...
+                    activeSuggestedCannedResponse: [['unlink']],
+                });
+            }
+        }
+
         /**
          * @private
          */
         _reset() {
+            this.closeCannedResponseSuggestions();
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:616]: `closeCannedResponseSuggestions` & `closeMentionSuggestions`. Latter should clarify partner mentions, but the anomaly is that both suggestions are treated too separately.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js [hunk 9]
```diff
@@ -270,11 +310,19 @@ class ComposerTextInput extends Component {
...
+                    } else {
+                        this.composer.setNextSuggestedCannedResponseActive();
+                    }
+                }
                 break;
             // Otherwise, check if a mention is typed
             default:
                 this.saveStateInStore();
                 this.composer._detectDelimiter();
+                this.composer._detectDelimiterCannedResponse();
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer_text_input/composer_text_input.js:325]: should be public method (also `_detectDelimited()` should have been public...)

**Pull request closed without merging**